### PR TITLE
tests: add dependency constraints to eslint-plugin tests

### PR DESCRIPTION
**Pull request by bradzacher**
## PR Checklist

- [x] Addresses an existing open issue: ref #1752
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Doing this ahead of #5573 to improve reviewability.
This PR marks all of our dependency requirements for tests back to TSv4.0.8 - I didn't go back further because #5915 bumps the minimum past that anyways

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/non-nullable-type-assertion-style.test.ts [hunk 1]
```diff
@@ -229,6 +197,54 @@ declare const x: T;
...
+    tsconfigRootDir: getFixturesRootDir(),
+    project: './tsconfig.noUncheckedIndexedAccess.json',
+  },
+  parser: '@typescript-eslint/parser',
+  dependencyConstraints: {
+    typescript: '4.1',
+  },
+});
+
+ruleTesterWithNoUncheckedIndexAccess.run('', rule, {
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/non-nullable-type-assertion-style.test.ts:215]:
```suggestion
ruleTesterWithNoUncheckedIndexAccess.run('non-nullable-type-assertion-style', rule, {
```

**Review by JoshuaKGoldberg — DISMISSED** [review]
🔥

**Review by JoshuaKGoldberg — DISMISSED** [review]

**Review by JoshuaKGoldberg — DISMISSED** [review]

**Comment by JoshuaKGoldberg** [comment]
I see some new test failures after my merge - will try to fix soon...

**Comment by JoshuaKGoldberg** [comment]
@bradzacher `non-nullable-type-assertion-style` has a failure - otherwise this looks good to go!

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
`non-nullable-type-assertion-style` has a failure - otherwise this looks good to go!

**Comment by bradzacher** [comment]
Interestingly the changes to that failing test have exposed a bug in our underlying project caching logic!
Fix is in #5999

**Review by JoshuaKGoldberg — APPROVED** [review]
🔥

**Pull request merged**